packages/java-edition/src/mcfunction/parser/argument.ts

```diff
@@ -1640,3 +1731,173 @@ function vector(
+				end: ',',
+				trailingEnd: true,
+			},
+			end: ']',
+		}),
+		(res) => {
+			const ans: ComponentListNode = {
+				type: 'mcfunction:component_list',
+				range: res.range,
+				children: res.children,
+			}
+			return ans
+		},
+	)
+}
+
+function componentTests(): core.InfallibleParser<ComponentPredicatesNode> {
+	const exact = core.map<
+		core.SequenceUtil<
+			core.ResourceLocationNode | core.LiteralNode | nbt.NbtNode
+		>,
+		ComponentTestExactNode
+	>(
+		core.sequence([
+			core.resourceLocation({
+				pool: [],
+				allowUnknown: true,
+				namespacePathSep: '.',
+			}),
+			core.literal('='),
+			nbt.parser.entry,
+		]),
+		(res) => {
+			const id = res.children.find(core.ResourceLocationNode.is)!
+			const value = res.children.find(nbt.NbtNode.is)!
+			const ans: ComponentTestExactNode = {
+				type: 'mcfunction:component_test_exact',
+				range: res.range,
+				children: [id, value],
+				component: id,
+				value,
+				negated: false,
+			}
+
+			return ans
+		},
+	)
+
+	const exists = core.map<
+		core.ResourceLocationNode,
+		ComponentTestExistsNode
+	>(
+		core.resourceLocation({
+			pool: [],
+			allowUnknown: true,
+			namespacePathSep: '.',
+		}),
+		(res) => {
+			const ans: ComponentTestExistsNode = {
+				type: 'mcfunction:component_test_exists',
+				range: res.range,
+				children: [res],
+				component: res,
+				negated: false,
+			}
+
+			return ans
+		},
+	)
+
+	const subpredicate = core.map<
+		core.SequenceUtil<
+			core.ResourceLocationNode | core.LiteralNode | nbt.NbtNode
+		>,
+		ComponentTestSubpredicateNode
+	>(
+		core.sequence([
+			core.resourceLocation({
+				pool: [],
+				allowUnknown: true,
+				namespacePathSep: '.',
+			}),
+			core.literal('~'),
+			nbt.parser.entry,
+		]),
+		(res) => {
+			const id = res.children.find(core.ResourceLocationNode.is)!
+			const subpredicate = res.children.find(nbt.NbtNode.is)!
+			const ans: ComponentTestSubpredicateNode = {
+				type: 'mcfunction:component_test_subpredicate',
+				range: res.range,
+				children: [id, subpredicate],
+				component: id,
+				subpredicate,
+				negated: false,
+			}
+
+			return ans
+		},
+	)
+
+	const not = core.map<
+		core.SequenceUtil<
+			| core.LiteralNode
+			| ComponentTestBaseNode
+		>,
+		ComponentTestBaseNode
+	>(
+		core.sequence([
+			core.literal('!'),
+			core.any([exact, subpredicate, exists]),
+		]),
+		(res) => {
+			const test = res.children.find(ComponentTestBaseNode.is)!
+			test.negated = true
+			return test
+		},
+	)
+
+	const test = core.select([
+		{
+			predicate: (src) => src.peek() === '!',
+			parser: not,
+		},
+		{
+			parser: core.map<
+				| ComponentTestExactNode
+				| ComponentTestExistsNode
+				| ComponentTestSubpredicateNode,
+				ComponentTestBaseNode
+			>(
+				core.any([exact, subpredicate, exists]),
+				(res) => res,
+			),
+		},
+	])
+
+	return core.map(
+		core.list({
+			start: '[',
+			value: test,
+			sep: [',', '|'],
+			end: ']',
+			trailingSep: true,
```

Comment:
This would allow trailing `|`'s, although upon my testing trailing commas aren't allowed by the game in item predicates (but they are allowed in non predicate), so just removing `trailingSep: true` probably works (would be nice if a comment could be added there to explain this). The parse tree also won't reflect the real hierarchy of the tests but I guess that doesn't matter much for us.

Reply:
The parse tree you see barely functions as is, unfortunately. There's (supposed to) be a push to fix it when I have the time